src/extensions/jana/JOmniFactory.h

```diff
@@ -58,10 +58,19 @@ class JOmniFactory : public JMultifactory {
   private:
     friend class JOmniFactory;
 
-    void GetCollection(const JEvent& event) { m_data = event.Get<T>(this->collection_names[0]); }
+    void GetCollection(const JEvent& event) {
+      try {
+        m_data = event.Get<T>(this->collection_names[0]);
+      } catch (const std::exception& e) {
```

Comment:
Can we narrow this to the specific exception when a collection is missing in the input and not the collection's algorithm made a boo-boo? I know, likely it's JException or bust, but I can dream, right?

Reply:
But this is just a rethrowing. I think motivation here was to provide nicer debug messages, not improve actual handling of errors.

Also, in principle, we could have added such context via boost::exceptions.

Reply:
Regarding narrowing by type, it's probably JException for collection not found.

Reply:
But we control the input layer through JEventSourcePodio. Does that help here? (On my phone so I'm not able to check in detail.)

Reply:
I don't what you have in mind here, we either have a type match or not in this clause. Generally, control over input layer doesn't help unless we can sort the dependency graph.

Reply:
Because JANA uses their "factory" pattern, dependencies turn into function call stacks. Any exception thrown in the deepest part of the call stack (frame 0 in a backtrace) is passed up and (hopefully*) rethrown without changing its type. It is up to the factory that allows for optional inputs (which is further up in the call stack) to decide what to do with the exception when it gets there. If we can throw a `jana::absent_in_event_source` then that exception is different from a `std::out_of_range` thrown somewhere.

* I think this is likely the part where it goes wrong. JANA takes a `std::exception` and throws a JException with only the `what()` string.

Reply:
Is there some advice with whether to do something different with the exception? This is something I haven't much experience with.

Reply:
I think I'm starting to get what you want here. I was also thinking about extending types of exceptions. But, in the end, it all requires that JANA throws them - there is a way to throw it from JEventSourcePODIO, I think.

Which of the following statuses are useful:
1. "not being present in input file"
2. "not being present in input file, but is an input that could have come from simulation"
3. "not being present in input file, but can not be produced by any factory available"

It looks like we'd like (2.), but (3.) is probably easier to derive.

Reply:
Should we move this further development to a future PR? Or is it likely if we don't tackle it now it will get ignored again for a long time?

Reply:
From my point of view this can get dealt with in a future PR, but I'd like to ask we print out at least a warning when a collection is ignored (until we can be sure it is ignored for the right reasons). And we should create an issue to keep track of this need.

Reply:
That would probably result in a lot of log messages for incomplete configurations, where every event could flag up every missing collection.

Looking at the capybara report again I am now confused as to how tracking_only already has entries in the ReconstructedParticles collection. Something about the behaviour must be different to what I've been assuming.

Reply:
Can we agree that exceptions arising from "missing" collections are limited to JException?
```suggestion
      } catch (const JException& e) {
```
This will leave more general class `std::exception`s to the status quo.

Reply:
> Looking at the capybara report again I am now confused as to how tracking_only already has entries in the ReconstructedParticles collection. Something about the behaviour must be different to what I've been assuming.

Do you know of a specific piece that should fail when calorimeters are missing?

Reply:
> > Looking at the capybara report again I am now confused as to how tracking_only already has entries in the ReconstructedParticles collection. Something about the behaviour must be different to what I've been assuming.
> 
> Do you know of a specific piece that should fail when calorimeters are missing?

I have done a couple of extra tests on the main branch which show I can recover the behaviour of the tracking_only configuration with the ip6 configuration.

The difference appears to be if I request `EcalClusters `or all three of `EcalEndcapPClusters,EcalBarrelScFiClusters,EcalEndcapNClusters` in my output_collections list, it forces the creation of the empty collection(s) even if the inputs don't exist, these empty collections then exist enough to propagate through the factories to fill the ReconstructedParticles collection.